api/classes/class-lightweight-api.php

```diff
@@ -0,0 +1,339 @@
+			if ( $db_host && $db_name && $db_user && $db_password && $db_charset ) {
+				$this->credentials = [
+					'db_host'     => $db_host,
+					'db_name'     => $db_name,
+					'db_user'     => $db_user,
+					'db_password' => $db_password,
+					'db_charset'  => $db_charset,
+					'db_prefix'   => $db_prefix,
+				];
+				return $this->credentials;
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Get transient name.
+	 */
+	public function get_transient_name() {
+		return sprintf( '%s-%s-popup', $this->client_id, $this->popup_id );
+	}
+
+	/**
+	 * Get suppression-related transient value
+	 *
+	 * @param string $prefix transient prefix.
+	 */
+	public function get_suppression_data_transient_name( $prefix ) {
+		return sprintf(
+			'%s-%s',
+			$prefix,
+			$this->client_id
+		);
+	}
+
+	/**
+	 * Get suppression-related transient value for newsletter-campaign-suppression, which has the order of client_id and key reversed from the others.
+	 *
+	 * @param string $prefix transient prefix.
+	 */
+	public function legacy_get_suppression_data_transient_name_reversed( $prefix ) {
+		return sprintf(
+			'%s-%s',
+			$this->client_id,
+			$prefix
+		);
+	}
+
+	/**
+	 * Complete the API and print response.
+	 */
+	public function respond() {
+		$this->debug['end_time'] = microtime( true );
+		$this->debug['duration'] = $this->debug['end_time'] - $this->debug['start_time'];
+		if ( defined( 'NEWSPACK_POPUPS_DEBUG' ) && NEWSPACK_POPUPS_DEBUG ) {
+			$this->response['debug'] = $this->debug;
+		}
+		http_response_code( 200 );
+		print json_encode( $this->response ); // phpcs:ignore
+		exit;
+	}
+
+	/**
+	 * Return a 500 code error.
+	 *
+	 * @param string $code The error code.
+	 */
+	public function error( $code ) {
+		http_response_code( 500 );
+		print json_encode( [ 'error' => $code ] ); // phpcs:ignore
+		exit;
+	}
+
+	/**
+	 * Get data from transient.
+	 *
+	 * @param string $name The transient's name.
+	 */
+	public function get_transient( $name ) {
+		global $wpdb;
+		$name = '_transient_' . $name;
+
+		$empty_transients = wp_cache_get( 'empty_transients', 'newspack-popups' );
+
+		if ( ! is_array( $empty_transients ) ) {
+			$empty_transients = array();
+		}
+
+		if ( isset( $empty_transients[ $name ] ) ) {
+			$this->debug['cache_count'] += 1;
+			return null;
+		}
+
+		$value = wp_cache_get( $name, 'newspack-popups' );
+		if ( false === $value ) {
+			$this->debug['read_query_count'] += 1;
+			$row  = $wpdb->get_row( $wpdb->prepare( "SELECT option_value FROM $wpdb->options WHERE option_name = %s LIMIT 1", $name ) ); // phpcs:ignore
+			if ( is_object( $row ) ) {
+				$value = $row->option_value;
+				wp_cache_add( $name, $value );
+			} else {
+				$empty_transients[ $name ] = true;
+				wp_cache_set( 'empty_transients', $empty_transients, 'newspack-popups' );
+			}
+		} else {
+			$this->debug['cache_count'] += 1;
+		}
+		return maybe_unserialize( $value );
+	}
+
+	/**
+	 * Upsert transient.
```

Comment:
I don't know what "Upsert" is even supposed to mean. :)

Reply:
I stand by it! https://docs.mongodb.com/manual/reference/method/Bulk.find.upsert/

Reply:
TIL!